cmd/snap-confine/snap-confine.c

```diff
@@ -90,6 +90,23 @@ int main(int argc, char **argv)
 #endif
 	struct sc_apparmor apparmor;
 	sc_init_apparmor_support(&apparmor);
+	if (!apparmor.is_confined && apparmor.mode != SC_AA_NOT_APPLICABLE
+	    && getuid() != 0 && geteuid() == 0) {
```

Comment:
To summarize this logic, only die() when:
- when snap-confine is compiled with apparmor support AND
- the system is capable of using apparmor AND
- the real uid is not root AND
- the effective uid is root

This should work to prevent attacks by malicious non-root, unconfined processes on the binary in the squashfs. It does not prevent against unconfined root processes attacking this binary, but that doesn't matter, an unconfined root process doesn't need this binary to escalate privileges.